internal/controller/pkg/transaction/reconciler.go

```diff
@@ -0,0 +1,284 @@
+	log := r.log.WithValues("request", req)
+
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	tx := &v1alpha1.Transaction{}
+	if err := r.kube.Get(ctx, req.NamespacedName, tx); err != nil {
+		log.Debug("cannot get Transaction", "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), "cannot get Transaction")
+	}
+
+	status := r.conditions.For(tx)
+
+	log = log.WithValues(
+		"uid", tx.GetUID(),
+		"version", tx.GetResourceVersion(),
+		"name", tx.GetName(),
+	)
+
+	if meta.WasDeleted(tx) {
+		return reconcile.Result{}, nil
+	}
+
+	if tx.IsComplete() {
+		log.Debug("Transaction is already complete")
+		return reconcile.Result{}, nil
+	}
+
+	limit := ptr.Deref(tx.Spec.RetryLimit, defaultRetryLimit)
+	if tx.Status.Failures >= limit {
+		log.Debug("Transaction failure limit reached", "limit", limit)
+
+		status.MarkConditions(xpv1.ReconcileSuccess(), v1alpha1.TransactionFailed(fmt.Sprintf("failure limit of %d reached", limit)))
+		return reconcile.Result{}, errors.Wrap(r.kube.Status().Update(ctx, tx), "cannot update Transaction status")
+	}
+
+	// Delete and Replace are not yet implemented. Fast-fail these transactions
+	// with a clear error message.
+	if tx.Spec.Change == v1alpha1.ChangeTypeDelete {
+		status.MarkConditions(xpv1.ReconcileSuccess(), v1alpha1.TransactionFailed("Delete transactions are not yet implemented"))
+		return reconcile.Result{}, errors.Wrap(r.kube.Status().Update(ctx, tx), "cannot update Transaction status")
+	}
+	if tx.Spec.Change == v1alpha1.ChangeTypeReplace {
+		status.MarkConditions(xpv1.ReconcileSuccess(), v1alpha1.TransactionFailed("Replace transactions are not yet implemented"))
+		return reconcile.Result{}, errors.Wrap(r.kube.Status().Update(ctx, tx), "cannot update Transaction status")
+	}
+
+	status.MarkConditions(v1alpha1.TransactionRunning())
+	if err := r.kube.Status().Update(ctx, tx); err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "cannot update Transaction status")
+	}
+
+	currentPackages, err := r.lock.Acquire(ctx, tx)
+	if errors.Is(err, ErrLockHeldByAnotherTransaction) {
+		log.Debug("Lock is held by another transaction")
+		status.MarkConditions(v1alpha1.TransactionBlocked("waiting for lock held by another transaction"))
+		return reconcile.Result{}, errors.Wrap(r.kube.Status().Update(ctx, tx), "cannot update Transaction status")
+	}
+	if err != nil {
+		log.Debug("cannot acquire lock", "error", err)
+		r.record.Event(tx, event.Warning(reasonLockAcquisition, errors.Wrap(err, "cannot acquire lock")))
+		tx.Status.Failures++
+		status.MarkConditions(xpv1.ReconcileError(errors.Wrap(err, "cannot acquire lock")))
+		_ = r.kube.Status().Update(ctx, tx)
+
+		return reconcile.Result{}, errors.Wrap(err, "cannot acquire lock")
+	}
+
+	// Always release lock before returning, even on success. Release is
+	// idempotent - if we committed, the annotation is already gone so this
+	// is a no-op.
+	defer func() {
+		_ = r.lock.Release(ctx, tx)
+	}()
```

Comment:
Does this imply that transactions can run/commit out of order? I.e., if transaction 2 fails, transaction 3 may run and succeed, but then transaction 2 can run and succeed on a subsequent retry?

Seems like a problem that we don't process/commit transactions in the order in which they're created. What if transaction 2 installs a package and transaction 3 deletes it? If transaction 2 fails transiently, could transaction 3 (vacuously) succeed, then transaction 2 succeed on retry, leaving the package installed?